### PR TITLE
CNV-13187: Create new fedora-realtime container disk for realtime testing

### PR DESCRIPTION
**Pull request by jordigilh**
This PR is part of the outgoing effort to expose the libvirt XML elements for [reduced CPU latency in realtime workloads](https://github.com/kubevirt/kubevirt/pull/6182).  

The objective of this PR is to create the scripts to build and publish a containerized image of Fedora 32 with the RT packages installed (kernel-rt and tuned-profiles-realtime). It also configures the tuned profile to use cores 0,1,2 for realtime and isolate managed IRQs.
```bash
isolated_cores=0,1,2
isolate_managed_irq=Y
```

/cc @dhiller 
/cc @vladikr @davidvossel

### CODE REVIEW

**Comment by vladikr** [comment]
:+1:

**Comment by vladikr** [comment]
@jordigilh are you planning to use this image as a kubevirtCI node or are you planning to use it as a guest OS image for testing real-time guests?

**Comment by jordigilh** [comment]
> @jordigilh are you planning to use this image as a kubevirtCI node or are you planning to use it as a guest OS image for testing real-time guests?

It's for the guest OS only.

File: cluster-provision/images/centos8-realtime/build.sh
Comment (dhiller) [inline cluster-provision/images/centos8-realtime/build.sh:15]: IMHO it might be cleaner to use an ARG in a static Dockerfile. This way we wouldn't have to clean up the uncommited changes after we run the script. WDYT?

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (dhiller) [inline cluster-provision/images/centos8-realtime/publish.sh:3]: I remember that quay will "garbage collect" untagged images after a while. Is that what we want? Otherwise we might consider using a tag like we do [in project-infra](https://github.com/kubevirt/project-infra/blob/main/images/publish_image.sh#L59). WDYT?

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (dhiller) [inline cluster-provision/images/centos8-realtime/publish.sh:4]: I've created the repository https://quay.io/repository/kubevirtci/centos8-realtime-container-disk and added you as maintainer.

Comment (jordigilh) [reply cluster-provision/images/centos8-realtime/publish.sh:4]: thanks!

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (jordigilh) [reply cluster-provision/images/centos8-realtime/publish.sh:3]: Sounds good. This image is meant to be used only for the CI/CD job, not for public consumption, so it makes sense to tag it specifically :smile:

File: cluster-provision/images/centos8-realtime/build.sh
Comment (jordigilh) [reply cluster-provision/images/centos8-realtime/build.sh:15]: Sounds good. I'll add a static Dockerfile with the ARG parameter. What do you think about removing the qcow2 image as part of the script cleanup?

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (jordigilh) [reply cluster-provision/images/centos8-realtime/publish.sh]: @dhiller changed and reused the `publish_image.sh` from project-infra since it has a lot of goodies that could become handy for any other image being generated here.
Let me know what you think :smile:

**Review by davidvossel — COMMENTED** [review]
I'm curious why this wasn't added to the`cluster-provision/images/vm-image-builder` directory.

I think you'd just need to create a subdirectory there similar to the `fedora-with-test-tooling` directory and put the custom logic related to the image and setup in there. Then you could reuse the existing build and publish scripts

Also, we've been hesitant to add more container disk images to kubevirt's functional test simply due to how large the images are to cache. It's not a huge deal to add another one if it's really needed, but in general we've been trying to combine as much as possible into the `fedora-with-test-tooling` image

Is it possible to expand on the `fedora-with-test-tooling` and make it usable for real time as well?

**Comment by jordigilh** [comment]
> I'm curious why this wasn't added to the`cluster-provision/images/vm-image-builder` directory.
> 
> I think you'd just need to create a subdirectory there similar to the `fedora-with-test-tooling` directory and put the custom logic related to the image and setup in there. Then you could reuse the existing build and publish scripts
> 
> Also, we've been hesitant to add more container disk images to kubevirt's functional test simply due to how large the images are to cache. It's not a huge deal to add another one if it's really needed, but in general we've been trying to combine as much as possible into the `fedora-with-test-tooling` image
> 
> Is it possible to expand on the `fedora-with-test-tooling` and make it usable for real time as well?

I think so, I'd have to install the real time packages on it. What I'm not sure is how to specify which kernel to boot from, because we'd have both time sliced and realtime kernels available. Do you know if it's possible and how?
In the meanwhile I've redone the centos-stream8 image using the vm-image-builder script.

**Comment by vladikr** [comment]
> I think so, I'd have to install the real time packages on it. What I'm not sure is how to specify which kernel to boot from, because we'd have both time sliced and realtime kernels available. Do you know if it's possible and how?

hmm. perhaps we can use the fedora image but has a separate container disk just for the realtime kernel.
In that case, we could use it for boot. See: https://github.com/kubevirt/kubevirt/pull/5416
What do you think?

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (dhiller) [reply cluster-provision/images/centos8-realtime/publish.sh]: You could also add `project-infra` as `extra_ref` as in this job: https://github.com/kubevirt/project-infra/blob/main/github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml#L13
Docs here: https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md
This way you could use the same script and would avoid duplication.

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (acardace) [inline cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:39]: Shouldn't you use `http://ccrma.stanford.edu/planetccrma/mirror/fedora/linux/planetccrma/32/x86_64/planetccrma-repo-1.1-3.fc32.ccrma.noarch.rpm` which is built for Fedora 32?

The rpm you're referencing is built for Fedora 16.

**Review by acardace — CHANGES_REQUESTED** [review]
Hi @jordigilh! Everything looks good to me but the single comment I've made.
And I guess you should amend your commit message to reflect the actual changes of the patch.

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (ormergi) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:39]: What is this package? why is it needed?

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (ormergi) [inline cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:41]: I am not sure if its going to actually take affect because it requires a reboot
Maybe you can workaround this by intervene during the script on the part it boots the VM with virt-install
and restart the VM manually.

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (ormergi) [inline cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:42]: I think its possible to `sysctl` to achieve that, isnt it?

**Review by ormergi — COMMENTED** [review]
Hi Jordi, 
I am not sure I fully understand how is it going to work,
the default kernel that will be used is the regular one (time-sliced?), and from the test suite we are going to boot the VM with the real-time kernel?

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (jordigilh) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:39]: @acardace Thanks! I will update.
@ormergi  This package installs the repo that contains the kernel-rt and other real time packages. I want to use the existing fedora-with-testing-tools for the real time functional tests. If this change has no impact to non-real time functional tests that leverage on this image, then we will avoid having to create a new container disk image just for real time functional tests.

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (jordigilh) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:41]: It does a shutdown at the end of the script. I tested this image locally and when it  boots up, it does with the real time kernel:
```
[fedora@vmi-fedora ~]$ uname -r
5.6.19-300.rt10.2.fc32.ccrma.x86_64+rt
```

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (jordigilh) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:42]: > Hi Jordi,
> I am not sure I fully understand how is it going to work,
> the default kernel that will be used is the regular one (time-sliced?), and from the test suite we are going to boot the VM with the real-time kernel?

The default kernel will be the real time kernel. What I need to test (in progress) is to validate that using the real time kernel won't impact the usage of this container disk in the functional tests. It's the alternative to having to create a new container disk just for real time tests and having to manage it.

**Comment by jordigilh** [comment]
> Hi @jordigilh! Everything looks good to me but the single comment I've made.
> And I guess you should amend your commit message to reflect the actual changes of the patch.

Thanks!. I plan to update the PR description and commit message once I have validated that the changes in the kernel won't impact the functional tests. I'll label this PR on hold until then, just to be sure it's not merged until validated.

**Comment by jordigilh** [comment]
I'm putting this PR on hold until I can confirm that using the real time kernel won't impact the functional tests in this other [PR](https://github.com/kubevirt/kubevirt/pull/6182) in the kubevirt project.
/hold

File: cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config
Comment (jordigilh) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:42]: > I think its possible to use `sysctl` to achieve that, isnt it?

I don't know, I need to check :smile:

Comment (jordigilh) [reply cluster-provision/images/vm-image-builder/fedora-with-test-tooling/cloud-config:42]: @ormergi it's not currently possible to change the settings with `sysctl`. As I understood from the feedback I've received, it's a work in progress.  The only other way is to inject these values as part of the kernel boot parameters.

**Comment by jordigilh** [comment]
> > > Another approach that I'm thinking is to default to kernel RT and only enable the tuned profile when running in real time. It's a simpler approach but it would need to be validated functional wise with the other usages of the fedora image. WDYT?
> > 
> > 
> > oh interesting. yeah that could work. I don't want you to feel like you're getting sucked to far down a rabbit hole here when all you want is to enable a functional test.
> > Here's what I suggest. Try making the fedora tools container disk default to using the RT kernel and see how CI behaves. If it looks like that's an option, then great. If CI falls apart, then don't spend a bunch of time trying to debug it, we'll just take in a new RT container disk based on either fedora or centos (which ever you'd like)
> 
> Terrific smile . I'll update the PR to add the `kernel-rt` packages for fedora.

@davidvossel it didn't work: one of the SRIOV tests that deals with live migration keeps [failing](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6182/pull-kubevirt-e2e-kind-1.19-sriov/1440698150317723648) when running with the realtime kernel. I've reached out to the network team but it seems the key people are on PTO and will be back next week (29SEP21). I propose to fork the fedora tooling image and create a new one with the realtime kernel enabled, so that it does not impact the SRIOV tests and allows the RT effort to be completed. What do you think?

**Comment by davidvossel** [comment]
> I propose to fork the fedora tooling image and create a new one with the realtime kernel enabled, so that it does not impact the SRIOV tests and allows the RT effort to be completed. What do you think?

Yep, that's the best we can do now. Thanks for looking into this

File: cluster-provision/images/centos8-realtime/publish.sh
Comment (jordigilh) [reply cluster-provision/images/centos8-realtime/publish.sh]: > I've created the repository https://quay.io/repository/kubevirtci/centos8-realtime-container-disk and added you as maintainer.

@dhiller , can you rename the repository to `https://quay.io/repository/kubevirtci/fedora-realtime-container-disk`? I ended up using Fedora and I got it working fine. Thanks!

**Review by dhiller — APPROVED** [review]
/approve

Thanks for your work!

**Pull request merged**
